File: win-prog/ProgWinVB/Pages and Transforms/ArbitraryCoordinates/ArbitraryCoordinates.vb
' --------------------------------------------------
' ArbitraryCoordinates.vb (c) 2002 by Charles Petzold
' --------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class ArbitraryCoordinates
    Inherits PrintableForm

    Shared Shadows Sub Main()
        Application.Run(New ArbitraryCoordinates())
    End Sub

    Sub New()
        Text = "Arbitrary Coordinates"
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        grfx.PageUnit = GraphicsUnit.Pixel
        Dim szf As SizeF = grfx.VisibleClipBounds.Size

        grfx.PageUnit = GraphicsUnit.Inch
        grfx.PageScale = Math.Min(szf.Width / grfx.DpiX / 1000, _
                                  szf.Height / grfx.DpiY / 1000)

        grfx.DrawEllipse(New Pen(clr), 0, 0, 990, 990)
    End Sub
End Class

File: win-prog/ProgWinCS/Pages and Transforms/ArbitraryCoordinates/ArbitraryCoordinates.cs
// --------------------------------------------------
// ArbitraryCoordinates.cs © 2001 by Charles Petzold
// --------------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class ArbitraryCoordinates: PrintableForm
{
     public new static void Main()
     {
          Application.Run(new ArbitraryCoordinates());
     }
     public ArbitraryCoordinates()
     {
          Text = "Arbitrary Coordinates";
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          grfx.PageUnit = GraphicsUnit.Pixel;
          SizeF sizef = grfx.VisibleClipBounds.Size;

          grfx.PageUnit  = GraphicsUnit.Inch;
          grfx.PageScale = Math.Min(sizef.Width  / grfx.DpiX / 1000, 
                                    sizef.Height / grfx.DpiY / 1000);

          grfx.DrawEllipse(new Pen(clr), 0, 0, 990, 990);
     }
}
